### ISSUE TITLE
there are competing implementations of a full subcategory

### ISSUE THREAD

**Issue by rmatthes** [UniMath/UniMath#1770]
The older one is found at https://github.com/UniMath/UniMath/blob/master/UniMath/CategoryTheory/Subcategory/Full.v#L71.
The one using displayed stuff is found at https://github.com/UniMath/UniMath/blob/master/UniMath/CategoryTheory/DisplayedCats/Constructions.v#L92.

The former file is imported by more than 20 files of the UniMath library. The newer definition tends to be used for constructions of monoidal categories. During the discussion of PR #1768, the incompatibility showed up.

**Comment by rmatthes** [comment]
I have the impression that my switch from the former to the latter implementation in the Eilenberg-Moore category now leads to unbearably slow compilation of PR #1768. This occurs in UniMath/CategoryTheory/EnrichedCats/Examples/EilenbergMooreEnriched.v. Did this file take 25min of compilation time already in the past? The biggest problems I see are in `nat_trans_to_eilenberg_moore_cat_enrichment` for rewrite steps.

**Comment by m-lindgren** [comment]
Is it a problem having two different implementations in the library? (besides the fact that the displayed one seems to cause compilation issues currently) Or are they not different enough to justify keeping both?

**Comment by rmatthes** [comment]
Good questions. The historic version is used in many places and certainly entered publications. The new version is needed to get the full monoidal subcategories. And that is used for the symmetric monoidal Eilenberg-Moore construction. But the base Eilenberg-Moore construction is with the historic version and needs to be some for the extension to the enriched setting.

**Comment by benediktahrens** [comment]
By '"needed", do you mean "needed" or "used"?

**Comment by nmvdw** [comment]
> I have the impression that my switch from the former to the latter implementation in the Eilenberg-Moore category now leads to unbearably slow compilation of PR #1768. This occurs in UniMath/CategoryTheory/EnrichedCats/Examples/EilenbergMooreEnriched.v. Did this file take 25min of compilation time already in the past? The biggest problems I see are in `nat_trans_to_eilenberg_moore_cat_enrichment` for rewrite steps.

That file didn't take 25 minutes in the past.

**Comment by m-lindgren** [comment]
So this is probably not related to the issue at hand, but this confuses me: Shouldn't the displayed implementation of full subcategories take a `P : hsubtype C` instead of `P : C -> UU`?

**Comment by rmatthes** [comment]
@m-lindgren : I agree that the condition for the displayed full subcategory to be univalent should rather be a prerequisite for the definition. That univalence of its total category needs the extra requirement that the original category is univalent is a different kind of requirement. Some years ago, definitions in UniMath were done with minimal requirements for being able to make them. But we have changed the policy. The original authors should speak up here.

**Comment by nmvdw** [comment]
We could implement full subcategories in the following ways:
1. The input is a category `C` and a predicate `P : C -> hProp` (i.e. `hsubtype C`)
2. The input is a category `C`, a type family `P : C -> UU` such that for all `x : C` we have `isaprop (P x)`
3. The input is a category `C` and a type family `P : C -> UU`. For the proof of univalence, one needs provide a proof that for all `x : C` we have `isaprop (P x)`

In addition, we can manage the assumptions differently for intermediate definitions (i.e., the displayed category) and for the final definition on univalent categories.

Of these three definitions, I prefer (2) over (1), because I think that (2) will be more convenient to use than (1). It is convenient to use `hProp` in settings where the predicates that you want to consider, are valued in `hProp`. However, if this is not the case, it can be more tedious to use `hProp`, because one needs to add a separate definition of the required `hProp`. Concretely, to define a full subcategory, you first add a helper definition which is the predicate for which you want to take the full subcategory.

These are (at least) two points where using `hProp` can cause annoyances:
- `is_z_isomorphism` is valued in `UU` and not in `hProp`. 
- `precategory_morphisms` is valued in `UU` and not in `hSet`. Because of this, one is not able to use the `logic` scope in order to get the equality type of morphisms as an `hProp`, and thus the following definition fails:
```coq
Fail Definition test
           (C : category)
           (x y : C)
           (f g : x --> y)
  : hProp
  := (f = g)%logic.
```
with the error `The term "f" has type "C ⟦ x, y ⟧" while it is expected to have type "pr1hSet ?X".`.

For this reason, I would avoid using either `hsubtype` or `C -> hProp`.

Note that there are ways to deal with these issues. The second one is dealt with by letting the notation `-->` and `C ⟦ x, y ⟧` only be defined for categories, and it should refer to `category_morphisms` whose type should be `hSet`. For the first one, there are a couple of methods:
- Add a definition `is_z_isomorphism_hProp`, which is valued in `hProp` that reduces to `is_z_isomorphism`. The disadvantage of this solution, is that this might be confusing to use.
- Prove that being an isomorphism is a proposition before defining `is_z_isomorphism`. This statement would then look something like `isaprop (∑ (inv : y --> x), ...)`. When defining `is_z_isomorphism`, its type should be `hProp`.

I also have a slight preference for 3 over 2 for the intermediate definitions for technical reasons. I would say that it would the nicest to define all relevant parts of the full displayed category in one section, so that the arguments only need to be stated once. However, the section mechanism disregards arguments that are not used in the definition in the section. As such, if one would do the following:
```coq
Section FullSub.
  Context (C : category)
                (P : C -> UU)
                (HP : ∏ (x : C), isaprop (P x)).
```
then the assumption `HP` would not be an argument in any definition in the section `FullSub` that does not use `HP`. This would mean that only the proof of univalence of the full category would have `HP` as an argument. (Note that this argument would not be valid in the first option).

Even without the requirement that `P` is not a predicate, I think this construction still warrants the name displayed full subcategory. For me, full subcategories are described by fully faithful functors, and even if you don't require `P` to be a predicate, you get this fully faithful functor. The only thing you lose by not requiring `P` to be a predicate, is the univalence of the full subcategory. I think that it would still be okay to call that `full_subcategory`.

For the reasons mentioned above, my preference would be as follows:
- For the intermediate definitions (the displayed category); the only arguments are those that are needed for the definition
- There is a definition `univalent_full_subcategory` with the following type signature
```coq
Definition univalent_full_sub
           {C : univalent_category}
           (P : C → UU)
           (HP : ∏ (x : C), isaprop (P x))
  : univalent_category.
```
